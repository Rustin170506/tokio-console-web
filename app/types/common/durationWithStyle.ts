import type { Duration } from "./duration";

/**
 * DurationWithStyle is a wrapper around the Duration type that includes a tailwind class to use for this duration.
 * It is used to display the duration in a color that represents the magnitude of the duration.
 */
export class DurationWithStyle {
    value: Duration;
    // The tailwind class to use for this duration.
    class: string;

    constructor(value: Duration, className: string) {
        this.value = value;
        this.class = className;
    }

    /**
     * Provides a human-readable string representation of the duration.
     * @returns a string representation of the duration.
     */
    toString(): string {
        return this.value.toString();
    }

    /**
     * Returns the duration as a number of microseconds.
     * Used for sorting.
     * @returns the duration as a number of microseconds.
     */
    valueOf(): number {
        return this.value.valueOf();
    }
}

export function getDurationWithClass(duration: Duration): DurationWithStyle {
    const days = duration.asDays();
    const hours = duration.asHours();
    const minutes = duration.asMinutes();
    const seconds = duration.asSeconds();
    const milliseconds = duration.asMilliseconds();

    let className: string;

    if (days >= 1) {
        className = "text-blue-500 dark:text-blue-300";
    } else if (hours >= 1) {
        className = "text-cyan-500 dark:text-cyan-300";
    } else if (minutes >= 1) {
        className = "text-green-500 dark:text-green-300";
    } else if (seconds >= 1) {
        className = "text-yellow-500 dark:text-yellow-300";
    } else if (milliseconds >= 1) {
        className = "text-red-500 dark:text-red-300";
    } else {
        className = "text-gray-500 dark:text-gray-300";
    }

    return new DurationWithStyle(duration, className);
}
