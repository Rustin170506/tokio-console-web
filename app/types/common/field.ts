import type { Metadata } from "./metadata";
import { Field as ProtoField } from "~/gen/common_pb";
import type { Location } from "~/gen/common_pb";

export function truncateRegistryPath(s: string): string {
    const regex = /.*\/cargo(\/registry\/src\/[^/]*\/|\/git\/checkouts\/)/;
    return s.replace(regex, "<cargo>/");
}

export function formatLocation(loc?: Location): string {
    if (loc) {
        let result = "";
        if (loc.modulePath) {
            result = loc.modulePath;
        } else if (loc.file) {
            const truncated = truncateRegistryPath(loc.file);
            result = truncated;
        } else {
            return "<unknown location>";
        }

        if (loc.line !== undefined) {
            result += `:${loc.line}`;

            if (loc.column !== undefined) {
                result += `:${loc.column}`;
            }
        }

        return result;
    }
    return "<unknown location>";
}

export enum FieldValueType {
    Bool = "Bool",
    Str = "Str",
    U64 = "U64",
    I64 = "I64",
    Debug = "Debug",
}

export class FieldValue {
    type: FieldValueType;
    value: boolean | string | bigint;

    constructor(type: FieldValueType, value: boolean | string | bigint) {
        this.type = type;
        this.value = value;
    }

    truncateRegistryPath(): FieldValue {
        if (
            (this.type === FieldValueType.Str ||
                this.type === FieldValueType.Debug) &&
            typeof this.value === "string"
        ) {
            this.value = truncateRegistryPath(this.value);
        }
        return this;
    }

    ensureNonempty(): FieldValue | undefined {
        if (
            (this.type === FieldValueType.Str ||
                this.type === FieldValueType.Debug) &&
            this.value === ""
        ) {
            return undefined;
        }
        return this;
    }

    static fromProto(field: ProtoField): FieldValue {
        switch (field.value.case) {
            case "debugVal":
                return new FieldValue(FieldValueType.Debug, field.value.value);
            case "strVal":
                return new FieldValue(FieldValueType.Str, field.value.value);
            case "u64Val":
                return new FieldValue(FieldValueType.U64, field.value.value);
            case "i64Val":
                return new FieldValue(FieldValueType.I64, field.value.value);
            case "boolVal":
                return new FieldValue(FieldValueType.Bool, field.value.value);
            default:
                throw new Error("Invalid field value case");
        }
    }
}

export class Field {
    static SPAWN_LOCATION = "spawn.location";
    static KIND = "kind";
    static NAME = "task.name";
    static TASK_ID = "task.id";

    name: string;
    value: FieldValue;

    constructor(name: string, value: FieldValue) {
        this.name = name;
        this.value = value;
    }

    static fromProto(field: ProtoField, meta: Metadata): Field | undefined {
        const nameValue = field.name;
        if (nameValue.case === undefined) {
            return undefined;
        }

        let name = "";
        switch (nameValue.case) {
            case "strName": {
                name = nameValue.value;
                break;
            }
            case "nameIdx": {
                const index = nameValue.value;
                const metaId = field.metadataId?.id;
                if (meta.id !== metaId) {
                    return undefined;
                }
                if (meta.fieldNames[Number(index)]) {
                    name = meta.fieldNames[Number(index)];
                } else {
                    return undefined;
                }
                break;
            }
        }

        let value = FieldValue.fromProto(field).ensureNonempty();
        // If the field is invalid or it has a string value which is empty, return null
        if (!name || !value) {
            return undefined;
        }

        // If the field name is 'SPAWN_LOCATION', truncate the registry path
        if (name === Field.SPAWN_LOCATION) {
            value = value.truncateRegistryPath();
        }

        return new Field(name.toString(), value);
    }
}
