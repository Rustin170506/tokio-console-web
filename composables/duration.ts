export class Duration {
    seconds: bigint;
    nanos: number;

    constructor(seconds: bigint, nanos: number) {
        this.seconds = seconds;
        this.nanos = nanos;
    }

    asDays(): number {
        return Number(this.seconds) / 86400;
    }

    asHours(): number {
        return Number(this.seconds) / 3600;
    }

    asMinutes(): number {
        return Number(this.seconds) / 60;
    }

    asSeconds(): number {
        return Number(this.seconds);
    }

    asMilliseconds(): number {
        return Number(this.seconds) * 1000 + this.nanos / 1e6;
    }

    asMicroseconds(): number {
        return Number(this.seconds) * 1e6 + this.nanos / 1000;
    }

    subtract(other: Duration): Duration {
        const result = new Duration(
            this.seconds - other.seconds,
            this.nanos - other.nanos,
        );
        return result;
    }

    add(other: Duration): Duration {
        const result = new Duration(
            this.seconds + other.seconds,
            this.nanos + other.nanos,
        );
        return result;
    }

    toString(): string {
        const totalSeconds = this.asSeconds();
        const units = [
            { value: Math.floor(this.asDays()), unit: "d" },
            { value: Math.floor(this.asHours()) % 24, unit: "h" },
            { value: Math.floor(this.asMinutes()) % 60, unit: "m" },
            { value: Math.floor(totalSeconds) % 60, unit: "s" },
            { value: Math.floor(this.asMilliseconds()) % 1000, unit: "ms" },
            { value: Math.floor(this.asMicroseconds()) % 1000, unit: "Âµs" },
            { value: this.nanos, unit: "ns" },
        ];

        let str = "";
        let partsIncluded = 0;

        if (totalSeconds >= 60) {
            // For durations longer than 60 seconds, include up to two non-zero units
            for (const { value, unit } of units) {
                if (value > 0) {
                    str += `${value}${unit} `;
                    partsIncluded++;
                    if (partsIncluded === 2) break;
                }
            }
        } else {
            // For durations shorter than 60 seconds, display the most significant non-zero unit
            for (const { value, unit } of units) {
                if (value > 0) {
                    str = `${value}${unit}`;
                    break;
                }
            }
        }

        return str || `${this.nanos}ns`;
    }
}
