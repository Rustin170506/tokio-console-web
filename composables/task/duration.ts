// The Duration class represents a duration of time.
export class Duration {
    // The number of seconds in the duration.
    seconds: bigint;
    // The number of nanoseconds in the duration.
    nanos: number;

    // The constructor takes the number of seconds and nanoseconds.
    constructor(seconds: bigint, nanos: number) {
        this.seconds = seconds;
        this.nanos = nanos;
        // Normalize the duration so that nanos is less than 1e9 and non-negative.
        this.normalize();
    }

    // Normalize ensures that nanos is less than 1e9 and non-negative.
    normalize(): Duration {
        let adjustment = 0;
        // If nanos is 1e9 or more, we need to adjust seconds and nanos.
        if (this.nanos >= 1e9) {
            adjustment = Math.floor(this.nanos / 1e9);
            this.seconds += BigInt(adjustment);
            this.nanos -= adjustment * 1e9;
        } else if (this.nanos < 0) {
            // If nanos is negative, we need to adjust seconds and nanos.
            adjustment = Math.floor(-this.nanos / 1e9) + 1;
            this.seconds -= BigInt(adjustment);
            this.nanos += adjustment * 1e9;
        }
        return this;
    }

    asDays(): number {
        return Number(this.seconds) / 86400;
    }

    asHours(): number {
        return Number(this.seconds) / 3600;
    }

    asMinutes(): number {
        return Number(this.seconds) / 60;
    }

    asSeconds(): number {
        return Number(this.seconds);
    }

    asMilliseconds(): number {
        return Number(this.seconds) * 1000 + this.nanos / 1e6;
    }

    asMicroseconds(): number {
        return Number(this.seconds) * 1e6 + this.nanos / 1000;
    }

    // subtract returns a new Duration that represents the difference between this duration and another.
    subtract(other: Duration): Duration {
        const seconds = this.seconds - other.seconds;
        const nanos = this.nanos - other.nanos;
        const result = new Duration(seconds, nanos);
        return result.normalize();
    }

    // add returns a new Duration that represents the sum of this duration and another.
    add(other: Duration): Duration {
        const result = new Duration(
            this.seconds + other.seconds,
            this.nanos + other.nanos,
        );
        return result.normalize();
    }

    // greaterThan checks if this duration is greater than another.
    greaterThan(other: Duration): boolean {
        return (
            this.seconds > other.seconds ||
            (this.seconds === other.seconds && this.nanos > other.nanos)
        );
    }

    toString(): string {
        const totalSeconds = this.asSeconds();
        const units = [
            { value: Math.floor(this.asDays()), unit: "d" },
            { value: Math.floor(this.asHours()) % 24, unit: "h" },
            { value: Math.floor(this.asMinutes()) % 60, unit: "m" },
            { value: Math.floor(totalSeconds) % 60, unit: "s" },
            { value: Math.floor(this.asMilliseconds()) % 1000, unit: "ms" },
            { value: Math.floor(this.asMicroseconds()) % 1000, unit: "Âµs" },
            { value: this.nanos, unit: "ns" },
        ];

        let str = "";
        let partsIncluded = 0;

        if (totalSeconds >= 60) {
            // For durations longer than 60 seconds, include up to two non-zero units
            for (const { value, unit } of units) {
                if (value > 0) {
                    str += `${value}${unit} `;
                    partsIncluded++;
                    if (partsIncluded === 2) break;
                }
            }
        } else {
            // For durations shorter than 60 seconds, display the most significant non-zero unit
            for (const { value, unit } of units) {
                if (value > 0) {
                    str = `${value}${unit}`;
                    break;
                }
            }
        }

        return str || `${this.nanos}ns`;
    }

    // valueOf returns the duration as a number of microseconds.
    // Used for sorting.
    valueOf(): number {
        return this.asMicroseconds();
    }

    // now returns the current time as a Duration from the Unix epoch.
    // TODO: use more precise time source. For example, performance.now() in browsers.
    static now(): Duration {
        return new Duration(BigInt(Math.floor(Date.now() / 1000)), 0);
    }

    static fromNano(nanos: bigint): Duration {
        const seconds = nanos / BigInt(1e9);
        const remainderNanos = nanos % BigInt(1e9);
        // After normalization, we can safely cast to number.
        return new Duration(seconds, Number(remainderNanos));
    }
}

// The Duration class can also be used as a timestamp.
export { Duration as Timestamp };
