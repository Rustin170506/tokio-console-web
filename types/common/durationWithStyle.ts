import type { Duration } from "./duration";

export class DurationWithStyle {
    value: Duration;
    // The tailwind class to use for this duration.
    class: string;

    constructor(value: Duration, className: string) {
        this.value = value;
        this.class = className;
    }

    toString(): string {
        return this.value.toString();
    }

    valueOf(): number {
        return this.value.valueOf();
    }
}

export function getDurationWithClass(duration: Duration): DurationWithStyle {
    const days = duration.asDays();
    const hours = duration.asHours();
    const minutes = duration.asMinutes();
    const seconds = duration.asSeconds();
    const milliseconds = duration.asMilliseconds();

    let className: string;

    if (days >= 1) {
        className = "text-blue-500 dark:text-blue-300";
    } else if (hours >= 1) {
        className = "text-cyan-500 dark:text-cyan-300";
    } else if (minutes >= 1) {
        className = "text-green-500 dark:text-green-300";
    } else if (seconds >= 1) {
        className = "text-yellow-500 dark:text-yellow-300";
    } else if (milliseconds >= 1) {
        className = "text-red-500 dark:text-red-300";
    } else {
        className = "text-gray-500 dark:text-gray-300";
    }

    return new DurationWithStyle(duration, className);
}
