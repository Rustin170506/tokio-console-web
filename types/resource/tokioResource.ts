/* eslint-disable camelcase */
import type { Duration, Timestamp } from "../common/duration";
import type { TokioResourceStats } from "./tokioResourceStats";
import { Resource_Kind, Resource_Kind_Known } from "~/gen/resources_pb";

export enum Visibility {
    Public,
    Internal,
}

export class TokioResource {
    // The task's pretty (console-generated, sequential) task ID.
    //
    // This is NOT the `tracing::span::Id` for the task's tracing span on the
    // remote.
    id: bigint;
    // The `tracing::span::Id` on the remote process for this task's span.
    //
    // This is used when requesting a task details stream
    spanId: bigint;
    parent: string;
    parentId: string;
    metaId: bigint;
    kind: string;
    stats: TokioResourceStats;
    target: string;
    concreteType: string;
    location: string;
    visibility: Visibility;

    constructor(
        id: bigint,
        spanId: bigint,
        parent: string,
        parentId: string,
        metaId: bigint,
        kind: string,
        stats: TokioResourceStats,
        target: string,
        concreteType: string,
        location: string,
        visibility: Visibility,
    ) {
        this.id = id;
        this.spanId = spanId;
        this.parent = parent;
        this.parentId = parentId;
        this.metaId = metaId;
        this.kind = kind;
        this.stats = stats;
        this.target = target;
        this.concreteType = concreteType;
        this.location = location;
        this.visibility = visibility;
    }

    totalDuration(since: Timestamp): Duration {
        const duration = since.subtract(this.stats.createdAt);
        return this.stats.total ?? duration;
    }

    isDropped(): boolean {
        return this.stats.total !== undefined;
    }
}

export function kindFromProto(pb: Resource_Kind): string {
    if (!pb.kind) {
        throw new Error("a resource should have a kind field");
    }

    switch (pb.kind.case) {
        case "known":
            if (pb.kind.value === Resource_Kind_Known.TIMER) {
                return "Timer";
            } else {
                throw new Error(`failed to parse known kind from ${pb.kind}`);
            }
        case "other":
            return pb.kind.value;
        default:
            throw new Error(`Unknown kind: ${pb.kind}`);
    }
}
