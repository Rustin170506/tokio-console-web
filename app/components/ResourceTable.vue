<template>
    <UTable
        :columns="columns"
        :rows="resources"
        :sort="{ column: 'id', direction: 'desc' }"
        :loading="pending"
        @select="select"
    >
        <template #loading-state>
            <div class="flex justify-center items-center h-full">
                <TokioSvg
                    class="w-8 h-auto mr-2 fill-black dark:fill-white my-8 animate-spin"
                />
            </div>
        </template>
        <template #total-data="{ row }">
            <p :class="`${row.total.class} w-10`">
                {{ row.total.toString() }}
            </p>
        </template>
        <template #visibilityIcon-data="{ row }">
            <UIcon :name="row.visibilityIcon" dynamic />
        </template>
        <template #location-data="{ row }">
            <LocationPopover :location="row.location" />
        </template>
        <template #attributes-data="{ row }">
            <AttributesPopover :attributes="row.attributes" />
        </template>
    </UTable>
</template>

<script lang="ts" setup>
import { useResources } from "~/composables/resources";
import {
    toResourceTableItem,
    type ResourceTableItem,
} from "~/types/resourceTableItem";

const columns = [
    {
        key: "id",
        label: "ID",
        sortable: true,
    },
    {
        key: "parentId",
        label: "Parent",
        sortable: true,
    },
    {
        key: "kind",
        label: "Kind",
        sortable: true,
    },
    {
        key: "total",
        label: "Total",
        sortable: true,
    },
    {
        key: "target",
        label: "Target",
        sortable: true,
    },
    {
        key: "type",
        label: "Type",
        sortable: true,
    },
    {
        key: "visibilityIcon",
        label: "Visibility",
        sortable: true,
        class: "w-1",
    },
    {
        key: "location",
        label: "Location",
    },
    {
        key: "attributes",
        label: "Attributes",
    },
];
const router = useRouter();

const select = (row: ResourceTableItem) => {
    router.push(`/resources/${row.id}`);
};

const { pending, resourcesData, lastUpdatedAt } = useResources();
const resources = computed(() => {
    return Array.from(resourcesData.value?.values() ?? [])
        .map((resource) => toResourceTableItem(resource, lastUpdatedAt.value!))
        .filter(Boolean) as ResourceTableItem[];
});
</script>
