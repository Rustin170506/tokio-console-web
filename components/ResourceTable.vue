<template>
    <UTable
        v-model:sort="sort"
        :columns="columns"
        :rows="resources"
        :loading="pending"
        @select="select"
    >
        <template #loading-state>
            <div class="flex justify-center items-center h-full">
                <TokioSvg
                    class="w-8 h-auto mr-2 fill-black dark:fill-white my-8 animate-spin"
                />
            </div>
        </template>
        <template #total-data="{ row }">
            <p :class="`${row.total.class} w-10`">
                {{ row.total.toString() }}
            </p>
        </template>
        <template #visibilityIcon-data="{ row }">
            <UIcon :name="row.visibilityIcon" dynamic />
        </template>
        <template #location-data="{ row }">
            <LocationPopover :location="row.location" />
        </template>
        <template #attributes-data="{ row }">
            <AttributesPopover :attributes="row.attributes" />
        </template>
    </UTable>
</template>

<script lang="ts" setup>
import { useResources } from "~/composables/resources";
import {
    toResourceTableItem,
    type ResourceTableItem,
} from "~/types/resourceTableItem";

const columns = [
    {
        key: "id",
        label: "ID",
        sortable: true,
    },
    {
        key: "parentId",
        label: "Parent",
        sortable: true,
    },
    {
        key: "kind",
        label: "Kind",
        sortable: true,
    },
    {
        key: "total",
        label: "Total",
        sortable: true,
    },
    {
        key: "target",
        label: "Target",
        sortable: true,
    },
    {
        key: "type",
        label: "Type",
        sortable: true,
    },
    {
        key: "visibilityIcon",
        label: "Visibility",
        sortable: true,
        class: "w-1",
    },
    {
        key: "location",
        label: "Location",
    },
    {
        key: "attributes",
        label: "Attributes",
    },
];
const router = useRouter();

const select = (row: ResourceTableItem) => {
    router.push(`/resources/${row.id}`);
};

const { pending, resourcesData, lastUpdatedAt } = useResources();
const resources = computed(() => {
    const lastUpdated = lastUpdatedAt.value;
    if (!lastUpdated) return [];

    return Array.from(resourcesData.value?.values() ?? [])
        .map((resource) => toResourceTableItem(resource, lastUpdated))
        .filter(Boolean) as ResourceTableItem[];
});

const sort = ref({
    column: "id",
    direction: "desc" as const,
});
</script>
