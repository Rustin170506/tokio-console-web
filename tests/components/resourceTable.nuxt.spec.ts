import { expect, it, describe } from "vitest";
import { mountSuspended, mockNuxtImport } from "@nuxt/test-utils/runtime";
import ResourceTable from "~/components/ResourceTable.vue";
import { Visibility, TokioResource } from "~/types/resource/tokioResource";
import { Timestamp } from "~/types/common/duration";
import type { TokioResourceStats } from "~/types/resource/tokioResourceStats";
import { Field, FieldValue, FieldValueType } from "~/types/common/field";

mockNuxtImport("useResources", () => {
    return () => {
        const pending = ref<boolean>(false);
        const resourcesData = ref<Map<bigint, TokioResource>>(new Map());
        const stats: TokioResourceStats = {
            createdAt: new Timestamp(1000n, 0),
            droppedAt: new Timestamp(2000n, 0),
            total: new Timestamp(1000n, 0),
            attributes: [
                {
                    field: new Field(
                        "target",
                        new FieldValue(FieldValueType.Str, "tokio:task"),
                    ),
                    unit: "ms",
                },
            ],
        };
        resourcesData.value.set(
            1n,
            new TokioResource(
                1n,
                1n,
                "N/A",
                "N/A",
                1n,
                "Timer",
                stats,
                "target",
                "sleep",
                "app.rs/68:10",
                Visibility.Public,
            ),
        );

        const lastUpdatedAt = ref<Timestamp>(new Timestamp(1000n, 0));
        return { pending, resourcesData, lastUpdatedAt };
    };
});

describe("ResourceTable", () => {
    it("renders correctly", async () => {
        const component = await mountSuspended(ResourceTable);
        expect(component.element.querySelector("tbody")).toMatchSnapshot();
    });

    it("has id column sorted by default", async () => {
        const component = await mountSuspended(ResourceTable);
        const idHeader = component.find('th[aria-sort="descending"]');
        expect(idHeader.exists()).toBe(true);
        expect(idHeader.find("span").text()).toBe("ID");
    });
});
