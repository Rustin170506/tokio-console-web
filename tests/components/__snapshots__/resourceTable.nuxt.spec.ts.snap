// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceTable > ResourceTable View 1`] = `
<tbody
  class="divide-y divide-gray-200 dark:divide-gray-800"
>
  
  
  <tr
    class="hover:bg-gray-50 dark:hover:bg-gray-800/50 cursor-pointer bg-slate-50 dark:bg-slate-950 animate-pulse"
  >
    <!--v-if-->
    <!--v-if-->
    
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      1
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      N/A
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      Timer
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm text-green-500 dark:text-green-300"
    >
      
      <p
        class="text-green-500 dark:text-green-300 w-10"
      >
        16m 40s 
      </p>
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      target
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      sleep
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      <span
        aria-hidden="true"
        class="iconify i-heroicons:eye"
        dynamic=""
      />
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      <div
        class="mb-1"
      >
        app.rs/68:10
      </div>
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      <div
        class="mb-1"
      >
        <p>
          <span
            class="text-blue-500 font-bold dark:text-blue-300"
          >
            target
          </span>
          <span
            class="text-gray-500 dark:text-gray-400"
          >
            =
          </span>
          <span
            class="text-yellow-500 dark:text-yellow-300"
          >
            tokio:task
          </span>
          <span
            class="text-blue-500 dark:text-blue-300"
          >
            ms
          </span>
        </p>
      </div>
      
    </td>
    
  </tr>
  <!--v-if-->
  
  
</tbody>
`;

exports[`ResourceTable View 1`] = `
<tbody
  class="divide-y divide-gray-200 dark:divide-gray-800"
>
  
  
  <tr
    class="hover:bg-gray-50 dark:hover:bg-gray-800/50 cursor-pointer bg-slate-50 dark:bg-slate-950 animate-pulse"
  >
    <!--v-if-->
    <!--v-if-->
    
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      1
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      N/A
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      Timer
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm text-green-500 dark:text-green-300"
    >
      
      <p
        class="text-green-500 dark:text-green-300 w-10"
      >
        16m 40s 
      </p>
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      target
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      sleep
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      <span
        aria-hidden="true"
        class="iconify i-heroicons:eye"
        dynamic=""
      />
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      <div
        class="mb-1"
      >
        app.rs/68:10
      </div>
      
    </td>
    <td
      class="whitespace-nowrap px-4 py-4 text-gray-500 dark:text-gray-400 text-sm"
    >
      
      <div
        class="mb-1"
      >
        <p>
          <span
            class="text-blue-500 font-bold dark:text-blue-300"
          >
            target
          </span>
          <span
            class="text-gray-500 dark:text-gray-400"
          >
            =
          </span>
          <span
            class="text-yellow-500 dark:text-yellow-300"
          >
            tokio:task
          </span>
          <span
            class="text-blue-500 dark:text-blue-300"
          >
            ms
          </span>
        </p>
      </div>
      
    </td>
    
  </tr>
  <!--v-if-->
  
  
</tbody>
`;
